### PR TITLE
introduce redis_throttle

### PR DESCRIPTION
**Pull request by gquintard**

### CODE REVIEW

**Comment by gquintard** [comment]
@ThijsFeryn , maybe this will interest you

**Comment by ThijsFeryn** [comment]
> @ThijsFeryn , maybe this will interest you

@gquintard Nice trick. Love the fact that you did this in VCL rather than having to patch the Redis VMOD.

File: vcls/redis_throttle/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+is a certain request should be denied. However, the quotas are kept in memory, and aren't shared,
+which makes the logic hard to scale when you have more than a few Varnish nodes.
+
+Instead, we can usea `redis` database and [vmod_redis](https://github.com/carlosabalde/libvmod-redis)
+to keep track of the various quotas to enforce.
+
+To use this VCL:
+- install `vmod_redis`, for example using [easy_install](https://github.com/varnish/toolbox/tree/master/easy_install).
+- open `redis_throttle.vtc` and edit it in all the places commented with `# EDIT` 
+- copy `redis_throttle.vtc` in `/etc/varnish` on your server
```
Comment (anthosz) [inline vcls/redis_throttle/README.md:19]: vtc to be replaced by vcl

File: vcls/redis_throttle/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+- in your own VCL (probably `/etc/varnish/default.vcl`), add `include "redis_throttle.vtc"` near the top
```
Comment (anthosz) [inline vcls/redis_throttle/README.md:20]: same ^^

File: vcls/redis_throttle/redis_throttle.vcl [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import std;
+import redis;
+
+# EDIT initialize vmod_redis, telling it where your redis database is
+sub vcl_init {
+    new db = redis.db(location="127.0.01:6379");
```
Comment (anthosz) [inline vcls/redis_throttle/redis_throttle.vcl:6]: typo: 127.0.0.1 for example

**Comment by anthosz** [comment]
Hello all,

@gquintard following our discussion on discord, just benchmarked redis vs no redis and it fix all the throttle issue in clustered mode.

Throttle used in the context (no specific backend used): 
```
  if (req.method == "GET" && req.http.XXXXXXX ~ "^XXXXXXXXXX.*$") {
    db.command("EVAL");
    db.push("""redis.call("set", KEYS[1], 0, "NX", "EX", ARGV[1]); if (redis.call("incr", KEYS[1]) > tonumber(ARGV[2])) then return "denied" else return "ok" end ;""");
    db.push(1);
    db.push(req.http.BLAH);
    db.push(60);
    db.push(1000);
    db.execute();
    if (db.reply_is_string() && db.get_reply() != "ok") {
      return(synth(429));
    }
    else {
      return(synth(200));
    }
  }
```

Without redis:
```
  if (req.method == "GET" && req.http.XXXXXXX ~ "^XXXXXXXXXX.*$") {
    if (vsthrottle.is_denied(req.http.BLAH, 1000, 60s)) {
      return(synth(429));
    }
    else {
      return(synth(200));
    }
  }
```

Scenarios: 1000 users (ramp-up: 1s/hold-for: 5m)
# 2 Varnish servers/ no redis
image
image
Total requests: 253320
200: 10867,43 (throttle status not shared so excepted)
Average response time: 154ms

# 2 Varnish servers/ 1 redis instance (separate server)
image
image
On redis side: memory used: 1MB / CPU less 3% (2 CPU) / 2520 Ops max
Total requests: 263755
200: 5011,345 (throttle status shared so excepted)
Average response time: 154ms

**Comment by gquintard** [comment]
@anthosz, this seems to indeed work well:

```
sub vcl_synth {
    db.command("EVAL");
    db.push("""redis.call("set", KEYS[1], 0, "NX", "EX", ARGV[1]); return tonumber(ARGV[2]) - redis.call("incr", KEYS[1]);""");
    db.push(1);
    db.push(req.url);
    db.push(100); // your period, in seconds
    db.push(17);  // your threshold for the period
    db.execute();

    if (db.reply_is_integer() && db.get_integer_reply() >= 0) {
        synthetic("ok ");
    } else {
        set resp.status = 429;
        synthetic("denied ");
    }
    return(deliver);
}
```

question is: do you need the expiration time too?

**Comment by anthosz** [comment]
@gquintard indeed, can be a bonus :)

**Comment by anthosz** [comment]
I guess it seems to do the job:
```
sub vcl_deliver {
    if (db.reply_is_integer()) {
        set resp.http.remaining = db.get_integer_reply()
    }
}
sub vcl_synth {
    db.command("EVAL");
    db.push("""redis.call("set", KEYS[1], 0, "NX", "EX", ARGV[1]); return tonumber(ARGV[2]) - redis.call("incr", KEYS[1]);""");
    db.push(1);
    db.push(req.url);
    db.push(100); // your period, in seconds
    db.push(17);  // your threshold for the period
    db.execute();

    if (db.reply_is_integer() && db.get_integer_reply() < 0) {
        set resp.status = 429;
        synthetic("denied ");
    }
    return(deliver);
}
```
The only issue that for each request, the remaining result decrease of 2 instead 1.

**Comment by anthosz** [comment]
But varnishtop output is more/less good (between 500 -> 1100 by minutes for a throttle set to 10000):
image

**Comment by gquintard** [comment]
```
sub vcl_deliver {
    if (db.reply_is_integer()) {
        set resp.http.remaining = db.get_integer_reply()
    }
}
sub vcl_synth {
    db.command("EVAL");
    db.push("""redis.call("set", KEYS[1], 0, "NX", "EX", ARGV[1]); return tonumber(ARGV[2]) - redis.call("incr", KEYS[1]);""");
    db.push(1);
    db.push(req.url);
    db.push(100); // your period, in seconds
    db.push(17);  // your threshold for the period
    db.execute();

    if (db.reply_is_integer() && db.get_integer_reply() < 0) {
        set resp.status = 429;
        synthetic("denied ");
    }
    return(deliver);
}
```

that code looks a bit weird, your response is either going to go to `vcl_synth` or to `vcl_deliver`, so you can't use `db.reply_is_integer()` in the latter, unless you executed a call earlier in your VCL, which could explain the -2 instead of -1 ?

**Comment by anthosz** [comment]
My bad (the laziness of sanitize), the used code:
```
sub vcl_deliver {
  if (db.reply_is_integer()) {
    set resp.http.XXXXXXXXXXXX = db.get_integer_reply();
  }
}

sub vcl_recv {
  if (req.method == "GET" && req.http.XXXXXXX ~ "^XXXXXXXXXX.*$") {
    db.command("EVAL");
    db.push("""redis.call("set", KEYS[1], 0, "NX", "EX", ARGV[1]); return tonumber(ARGV[2]) - redis.call("incr", KEYS[1]);""");
    db.push(1);
    db.push(req.http.BLAH);
    db.push(60);
    db.push(1000);
    db.execute();
    if (db.reply_is_integer() && db.get_integer_reply() < 0) {
      return(synth(429));
    }
  }
}
```

**Comment by gquintard** [comment]
weird, I'm getting the proper decrement of 1 with that code :-/

**Comment by gquintard** [comment]
I'll merge this, @anthosz, let me know if you have any issue with this

**Comment by anthosz** [comment]
@gquintard Thank you! It's in production since few months and it works like a charm!

**Pull request merged**